src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+                    .mapNotNull { (queryOp, queryFieldName) ->
+                        val (index, objectField) = queriedFields.remove(queryFieldName) ?: return@mapNotNull null
+                        val op = when (definedField.type.isList()) {
+                            true -> when (queryOp) {
+                                RelationOperator.NOT -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.NOT
+                                    else -> RelationOperator.NONE
+                                }
+                                RelationOperator.EQ_OR_NOT_EXISTS -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.EQ_OR_NOT_EXISTS
+                                    else -> {
+                                        LOGGER.info("$queryFieldName on type ${type.name} was used for filtering, consider using ${definedField.name}${RelationOperator.EVERY.suffix} instead")
+                                        RelationOperator.EVERY
+                                    }
+                                }
+                                else -> queryOp
+                            }
+                            false -> when (queryOp) {
+                                RelationOperator.SINGLE -> {
+                                    LOGGER.warn("Using $queryFieldName on type ${type.name} is deprecated, use ${definedField.name} directly")
+                                    RelationOperator.SOME
+                                }
+                                RelationOperator.SOME -> {
+                                    LOGGER.warn("Using $queryFieldName on type ${type.name} is deprecated, use ${definedField.name} directly")
+                                    RelationOperator.SOME
+                                }
+                                RelationOperator.NONE -> {
+                                    LOGGER.warn("Using $queryFieldName on type ${type.name} is deprecated, use ${definedField.name}${RelationOperator.NOT.suffix} instead")
+                                    RelationOperator.NONE
+                                }
+                                RelationOperator.NOT -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.NOT
+                                    else -> RelationOperator.NONE
+                                }
+                                RelationOperator.EQ_OR_NOT_EXISTS -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.EQ_OR_NOT_EXISTS
+                                    else -> RelationOperator.SOME
+                                }
+                                else -> queryOp
+                            }
+                        }
+                        Predicate(op, objectField, definedField, index)
+                    }
+                    .forEach { relationPredicates.add(it) }
```

Comment:
again, is there a better way to do this without return@ and mutating the map here?
and without mutating relPredicates